include/odp/api/spec/packet_io_types.h

```diff
@@ -1086,6 +1092,18 @@ typedef struct odp_pktio_capability_t {
 
 } odp_pktio_capability_t;
 
+/** Parameters for ODP_LSO_WRITE_BITS custom operation */
+typedef struct odp_lso_write_bits_t {
+	/** Bitmask to select which bits to write */
+	uint8_t mask;
+
+	/** Value to be written using the mask:
+	 *  new_value := (old_value & ~mask) | (value & mask)
```

Comment:
Both value and mask are user arguments for the profile, so there’s no need for the extra & operation. We can change this to new_value := (old_value & ~mask) | value.

Reply:
This is just a matter of taste regarding how the API is defined without effect on the provided functionality. I think it is clearer that the mask field defines which bits are going to be affected so that other bits in the value field not have any effect. I do not see why your proposal would be preferable as opposed to just different.

Reply:
What if other bits in the value field need to be set irrespective of the old value?

Reply:
I think I may not understand the question. The bits selected by the mask will be set of cleared based on the value irrespective of the old value of those bits in the destination byte.

Reply:
Sorry for the confusion, my intention was as follows:

1) Mask Application: The bits specified by the mask will clear the corresponding bits in old_value and set them based on the same bit positions from value.

2) Remaining Bits: The bits in old_value that are not covered by the mask can either remain untouched or be selectively enabled based on value. Since the mask does not cover these bits, their behavior is implementation-defined. Therefore, the API specification will not explicitly state that these bits are untouched.

Reply:
I still do not quite get it. The API says that the mask selects which bits to write. So the bits not selected by the mask are not written to, so they retain their values. Also, the expression in the API text (new_value := (old_value & ~mask) | (value & mask)) clarifies what is going to happen. Thus I do not see any ambiguity here regarding the bits not set in the mask.

Reply:
Assuming bits not selected by the mask should always be 1 irrespective of old_value is also a valid case. Take the examples below:

Formula 1: new_value := (old_value & ~mask) | (value & mask)

Case 1:

old_value: 0b11001100
mask: 0b11110000
value: 0b10101010
new_value: 0b10101100

Case 2:

old_value: 0b11001100
mask: 0b11110000
value: 0b10100000
new_value: 0b10101100

Formula 2: new_value := (old_value & ~mask) | value

Case 1:

old_value: 0b11001100
mask: 0b11110000
value: 0b10101010
new_value: 0b10101110

Case 2:

old_value: 0b11001100
mask: 0b11110000
value: 0b10100000
new_value: 0b10101100

With your Formula 1, bits not set in the mask remain the same irrespective of value, but the user can select a suitable output based on different values as mentioned in Formula 2. I think either the API will not specify anything about bits not selected by the mask, or it should be generic like Formula 2.

Reply:
> Assuming bits not selected by the mask should always be 1 irrespective of old_value is also a valid case.

Why? If you want to set certain bit, then you can include that bit in the mask.


> With your Formula 1, bits not set in the mask remain the same irrespective of value

Yes, that is correct.

> but the user can select a suitable output based on different values as mentioned in Formula 2

Yes they could, but they could do the same by simply setting the necessary bits in the mask.

> I think either the API will not specify anything about bits not selected by the mask

That would not be a good idea.

> or it should be generic like Formula 2.

But why? What would be the benefit? Now the API specifies that the bits not set in the mask are not changed in the target byte. I think that is quite simple and clear.

Reply:
As per the spec, we have two things:

1) Functionality for “bits set in mask.” We are both on the same page here.
2) Functionality for “bits not set in the mask.” You are proposing the new value will be the same as the old value, whereas I am proposing the new value will be “old value ORed with the given value.” Given an example where Formula 2 works for both cases, whereas Formula 1 does not.
The implementation is simple for both. It is up to you to pick the formula that supports multiple cases without changing the mask, or you can stick to a single functionality.

Reply:
We seem to be going in circles.

The proposed API change in this PR makes it possible to indicate which packet bits are to be changed and in which way at segmentation time, don't you agree? The API also does not leave any ambiguity on the outcome of any valid configuration is. I still do not see what problem you see in the API.

> Given an example where Formula 2 works for both cases, whereas Formula 1 does not.

I really do not understand what you mean when you say that formula 1 "does not work". Can you elaborate? In which situation does it not work?

> Functionality for “bits not set in the mask.” You are proposing the new value will be the same as the old value, 

Correct. The mask specifies which bits are to be changed. The value specifies the new value of those bits. Value bits that are outside the masked bits are thus ignored. It seems that you have a problem with having those bits ignored, but I do not understand what the problem is, other than that you do not like it. If an API user does not want to have certain bit ignored, then he/she can simply include it in the mask. That is what the mask is for.

> It is up to you to pick the formula that supports multiple cases without changing the mask, or you can stick to a single functionality.

There is only single functionality here, not multiple ones. The single functionality is to set selected bits to selected values. That is all. The fact that your API alternative makes it possible to specify the desired outcome in two different ways does not mean that there are two functionalities that could be supported. It only means that your API has two ways to express the same thing, which some might see as a weakness rather than strength. I think the API is more clear when the mask indicates which bits and which bits only are to be written to.

Reply:
As discussed offline, both ways of expressing the bits-to-be-changed and their new values are essentially equivalent and express the same thing in slightly different way. So which one to choose is just a matter of taste and does not have any effect on what can or can not be done.